### PR TITLE
add tests for redis

### PR DESCRIPTION
**Pull request by madhur-tandon**
For some reason, the tests fail. But, do these changes suffice for adding the test? What about the `async` and `await` keywords? Are they applicable here? CC: @btel @wolfv 

Also, they won't run on CI for now. So the above comment was for failing locally.

### CODE REVIEW

**Comment by madhur-tandon** [comment]
Update: The tests pass locally, only the CI remains to be configured.

File: quetz/deps.py [hunk 1]
```diff
@@ -95,15 +95,9 @@ def get_tasks_worker(
...
+        worker = SubprocessWorker(auth.API_key, auth.session, config)
     elif worker == "redis-queue":
-        worker = RQManager(
-            config.redis_ip,
-            config.redis_port,
-            config.redis_db,
-            auth.API_key,
-            auth.session,
-        )
+        worker = RQManager(auth.API_key, auth.session, config)
```
Comment (btel) [inline quetz/deps.py:100]: why did you remove the ip/port configuration?

File: quetz/testing/fixtures.py [hunk 2]
```diff
@@ -144,6 +144,8 @@ def config_base(database_url, plugins):
 
 [plugins]
 enabled = {plugins}
+
+[redis]
```
Comment (btel) [inline quetz/testing/fixtures.py:148]: make sure that the tests also pass if [redis] section is not present, you can check that in the code using `config.configured_section("redis")`

File: quetz/tasks/workers.py [hunk 3]
```diff
@@ -29,7 +29,7 @@ def prepare_arguments(func: Callable, **resources):
     return kwargs
 
 
-def wrapper(func, api_key, browser_session, **kwargs):
+def wrapper(func, api_key, browser_session, config, **kwargs):
```
Comment (btel) [inline quetz/tasks/workers.py:32]: maybe more telling  function name and docstring would be helpful

File: quetz/testing/fixtures.py (hunk 2 above)
Comment (btel) [reply quetz/testing/fixtures.py:148]: can you also add the extra config settings to the sphinx docs, it can be done in a separate PR

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -30,6 +30,15 @@ jobs:
...
           --health-retries 5
+      redis:
+        image: redis
+        options: >-
+          --health-cmd "redis-cli ping"
+          --health-interval 10s
+          --health-timeout 5s
+          --health-retries 5
+        ports:
+          - 6379:6379
```
Comment (btel) [inline .github/workflows/ci.yml:41]: I don't see where you initialize the workers and the redis server, it can be done in the CI script or in the test

**Review by btel — COMMENTED** [review]
looks great! Just a few, hopefully helpful, comments, cheers

File: quetz/deps.py (hunk 1 above)
Comment (madhur-tandon) [reply quetz/deps.py:100]: because we are getting it from `config` now.

File: quetz/testing/fixtures.py (hunk 2 above)
Comment (madhur-tandon) [reply quetz/testing/fixtures.py:148]: yeah, thanks, I didn't know that we wanted to make it pass even without the `[redis]` section. Will do the docs too.

File: quetz/tasks/workers.py (hunk 3 above)
Comment (madhur-tandon) [reply quetz/tasks/workers.py:32]: cool, agreed. will change.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (madhur-tandon) [reply .github/workflows/ci.yml:41]: Yup, this was the last missing piece.

**Comment by btel** [comment]
it looks good, it would be also great to add docs about setting up quetz with redis workers (launching redis, rq and workers, + configuration that we already discussed)

File: docs/source/deploying/configuration.rst [hunk 5]
```diff
@@ -122,6 +122,24 @@ Quetz can store package in object cloud storage compatible with S3 interface. To
...
+``redis`` section
+^^^^^^^^^^^^^^^^^^^^^^
+
+Quetz can be configured to run with a redis-server and use redis-queue for queueing jobs and processing them in the background with workers.
+Note that you might need to run ``redis-server`` and ``rq worker`` commands in separate terminals / containers along with quetz.
+You can configure the redis-server by setting the host, port and db:
+
+.. code::
+
+   [sqlalchemy]
```
Comment (wolfv) [inline docs/source/deploying/configuration.rst:134]: `[redis]`?

Comment (madhur-tandon) [reply docs/source/deploying/configuration.rst]: yeah, fixed

File: docs/source/deploying/configuration.rst [hunk 6]
```diff
@@ -122,6 +122,24 @@ Quetz can store package in object cloud storage compatible with S3 interface. To
 :bucket_prefix:
 :bucket_suffix: channel directories on S3 are created with the following semantics: ``{bucket_prefix}{channel_name}{bucket_suffix}``
 
+``redis`` section
+^^^^^^^^^^^^^^^^^^^^^^
+
+Quetz can be configured to run with a redis-server and use redis-queue for queueing jobs and processing them in the background with workers.
+Note that you might need to run ``redis-server`` and ``rq worker`` commands in separate terminals / containers along with quetz.
```
Comment (btel) [inline docs/source/deploying/configuration.rst:129]: can you create a new file in docs/source/deploying/workers.rst describing the procedure step by step, including the installation and configuration of redis (you can also pull redis from docker).

the config options should be also given  here.

File: docs/source/deploying/configuration.rst [hunk 7]
```diff
@@ -122,6 +122,24 @@ Quetz can store package in object cloud storage compatible with S3 interface. To
...
+You can configure the redis-server by setting the host, port and db:
+
+.. code::
+
+   [redis]
+   ip = "127.0.0.1"
+   port = 6379
+   db = 0
```
Comment (btel) [inline docs/source/deploying/configuration.rst:137]: what is db? is this an integer or string? how can we make sure that the db exist?

File: docs/source/deploying/configuration.rst (hunk 6 above)
Comment (btel) [reply docs/source/deploying/configuration.rst:129]: you can also mention the other worker types. If you don't have time, I can also do this after this PR is merged.

File: docs/source/deploying/configuration.rst [hunk 8]
```diff
@@ -122,6 +122,24 @@ Quetz can store package in object cloud storage compatible with S3 interface. To
 :bucket_prefix:
 :bucket_suffix: channel directories on S3 are created with the following semantics: ``{bucket_prefix}{channel_name}{bucket_suffix}``
 
+``redis`` section
```
Comment (btel) [inline docs/source/deploying/configuration.rst:125]: I guess you will also need to document the `[workers]` section

File: docs/source/deploying/configuration.rst (hunk 6 above)
Comment (madhur-tandon) [reply docs/source/deploying/configuration.rst:129]: We should do this in a separate PR.

File: docs/source/deploying/configuration.rst (hunk 7 above)
Comment (madhur-tandon) [reply docs/source/deploying/configuration.rst]: The default database that `redis` uses is called `db` 0 to store stuff.

File: quetz/deps.py [hunk 9]
```diff
@@ -92,17 +92,22 @@ def get_tasks_worker(
     else:
         worker = "thread"
 
+    if config.configured_section("redis"):
+        redis_ip = config.redis_ip
+        redis_port = config.redis_port
+        redis_db = config.redis_db
+    else:
+        redis_ip = "127.0.0.1"
```
Comment (btel) [inline quetz/deps.py:100]: I am not sure whether these are reasonable defaults, perhaps we should force the user to provide the values if the worker is set to redis?

Comment (madhur-tandon) [reply quetz/deps.py:100]: Yeah, we can do that too. The default `IP`, `Port`, and `db` are always those if you run a server `locally`.

File: quetz/tasks/workers.py [hunk 10]
```diff
@@ -143,21 +151,38 @@ async def wait(self):
...
+            **kwargs,
         )
 
-    def wait(self):
+    async def wait(self):
         while not self.job.is_finished:
             time.sleep(1)
-        return self.job.result
+        if self.job.result:
+            return await self.job.result
```
Comment (btel) [inline quetz/tasks/workers.py:188]: you can't await results, it's not a asynchronous function, but the result of the execution (`None` in this case).

File: quetz/deps.py (hunk 9 above)
Comment (btel) [reply quetz/deps.py:100]: so maybe we can keep them, but the defaults should be handled in quetz.config.Config, not here

File: quetz/tasks/workers.py (hunk 10 above)
Comment (madhur-tandon) [reply quetz/tasks/workers.py:188]: This is done to make it compatible with the `tests`. So that the statement `await any_worker.wait()` in `test_threading_worker_execute` and `test_threading_worker_execute_with_dao` works for any type of worker. It anyway worked for `threading` and `subprocess`. But it fails with `redis` if `async` and `await` are not used.

Comment (madhur-tandon) [reply quetz/tasks/workers.py:188]: But I guess we can do away with `await` here. And we will still need to use `async` perhaps.

File: docs/source/deploying/configuration.rst (hunk 6 above)
Comment (madhur-tandon) [reply docs/source/deploying/configuration.rst]: Done via https://github.com/mamba-org/quetz/pull/202
